src/Akka.Streams.Kafka/Stages/Consumers/CommitCollectorStage.cs

```diff
@@ -0,0 +1,167 @@
+        {
+            var offset = Grab(Stage.In);
+            Log.Debug("Consuming offset {0}", offset);
+            if (ObservationLogic.UpdateBatch(offset))
+            {
+                // Push only of the outlet is available, a commit on interval might have taken the pending demand.
+                // This is very hard to get tested consistently, so it gets this big comment instead.
+                if (IsAvailable(Stage.Out))
+                {
+                    PushDownStream(CommitTrigger.BatchSize.Instance);
+                }
+                
+            }
+            else
+            {
+                TryPull(Stage.In);
+            }
+        }
+
+        private void OnPull()
+        {
+            if (_pushOnNextPull)
+            {
+                PushDownStream(CommitTrigger.Interval.Instance);
+                _pushOnNextPull = false;
+            }
+            else if (!HasBeenPulled(Stage.In))
+            {
+                TryPull(Stage.In);
+            }
+        }
+
+        public CommitObservationLogic ObservationLogic { get; }
+        public CommitCollectorStage Stage { get; }
+        public Attributes InheritedAttributes { get; }
+
+        public CommitterSettings Settings => Stage._settings;
+
+        public bool ActiveBatchInProgress => !ObservationLogic.OffsetBatch.IsEmpty;
+
+        protected override object LogSource => typeof(CommitCollectorStageLogic);
+
+        private bool _pushOnNextPull = false;
+
+        public override void PreStart()
+        {
+            base.PreStart();
+            ScheduleCommit();
+            Log.Debug("CommitCollectorStage initialized");
+        }
+
+        protected override void OnTimer(object timerKey)
+        {
+            switch (timerKey)
+            {
+                case CommitNow:
+                {
+                    if (ActiveBatchInProgress)
+                    {
+                        // Push only of the outlet is available, as timers may occur outside a push/pull cycle.
+                        // Otherwise instruct `OnPull` to emit what is there when the next pull occurs.
+                        // This is very hard to get tested consistently, so it gets this big comment instead.
+                        if (IsAvailable(Stage.Out))
+                        {
+                            PushDownStream(CommitTrigger.Interval.Instance);
+                        }
+                        else
+                        {
+                            _pushOnNextPull = true;
+                        }
+                    }
+                    else
+                    {
+                        ScheduleCommit();
+                    }
+
+                    break;
+                }
+                default:
+                    Log.Warning("Unexpected timer [{0}]", timerKey);
+                    break;
+            }
+        }
+
+        private void PushDownStream(CommitTrigger.ITriggeredBy triggeredBy)
```

Comment:
If we hit this method it's because we were triggered by batch size or by time delay - that gets logged here but otherwise we prepare the next batch and push the current one downstream.